File: mm-cs/src/MetaMessage/Mm/TagFieldParser.cs
namespace MetaMessage.Mm;

public static class TagFieldParser
{
    public static MmTag Parse(byte[] data)
    {
        MmTag tag = MmTag.Empty();
        int offset = 0;

        while (offset < data.Length)
        {
            if (offset >= data.Length)
                break;

            byte fieldType = data[offset++];

            switch (fieldType)
            {
                case 1: // Name
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    int nameLen = data[offset++];
                    if (offset + nameLen > data.Length)
                        throw new MmDecodeException("Tag name data overflow");
                    tag.Name = System.Text.Encoding.UTF8.GetString(data, offset, nameLen);
                    offset += nameLen;
                    break;

                case 2: // Desc
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    int descLen = data[offset++];
                    if (offset + descLen > data.Length)
                        throw new MmDecodeException("Tag desc data overflow");
                    tag.Desc = System.Text.Encoding.UTF8.GetString(data, offset, descLen);
                    offset += descLen;
                    break;

                case 3: // Type
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    tag.Type = (ValueType)data[offset++];
                    break;

                case 4: // ChildType
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    tag.ChildType = (ValueType)data[offset++];
                    break;

                case 5: // Nullable
                    tag.Nullable = true;
                    break;

                case 6: // Raw
                    tag.Raw = true;
                    break;

                case 7: // AllowEmpty
                    tag.AllowEmpty = true;
                    break;

                case 8: // Unique
                    tag.Unique = true;
                    break;

                case 9: // DefaultValue
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    int defaultValueLen = data[offset++];
                    if (offset + defaultValueLen > data.Length)
                        throw new MmDecodeException("Tag default value data overflow");
                    tag.DefaultValue = System.Text.Encoding.UTF8.GetString(data, offset, defaultValueLen);
                    offset += defaultValueLen;
                    break;

                case 10: // EnumValues
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    int enumCount = data[offset++];
                    tag.EnumValues = new List<string>();
                    for (int i = 0; i < enumCount; i++)
                    {
                        if (offset >= data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        int enumValueLen = data[offset++];
                        if (offset + enumValueLen > data.Length)
                            throw new MmDecodeException("Tag enum value data overflow");
                        tag.EnumValues.Add(System.Text.Encoding.UTF8.GetString(data, offset, enumValueLen));
                        offset += enumValueLen;
                    }
                    break;

                case 11: // LocationHours
                    if (offset + 3 >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    tag.LocationHours = (data[offset] << 24) | (data[offset + 1] << 16) | (data[offset + 2] << 8) | data[offset + 3];
                    offset += 4;
                    break;

                case 12: // Version
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    tag.Version = data[offset++];
                    break;

                case 13: // Mime
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    int mimeLen = data[offset++];
                    if (offset + mimeLen > data.Length)
                        throw new MmDecodeException("Tag mime data overflow");
                    tag.Mime = System.Text.Encoding.UTF8.GetString(data, offset, mimeLen);
                    offset += mimeLen;
                    break;

                case 14: // ChildDesc
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    int childDescLen = data[offset++];
                    if (offset + childDescLen > data.Length)
                        throw new MmDecodeException("Tag child desc data overflow");
                    tag.ChildDesc = System.Text.Encoding.UTF8.GetString(data, offset, childDescLen);
                    offset += childDescLen;
                    break;

                case 15: // ChildNullable
                    tag.ChildNullable = true;
                    break;

                case 16: // ChildEnum
                    if (offset >= data.Length)
                        throw new MmDecodeException("Unexpected end of tag data");
                    int childEnumCount = data[offset++];
                    tag.ChildEnum = new List<string>();
                    for (int i = 0; i < childEnumCount; i++)
                    {
                        if (offset >= data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        int childEnumValueLen = data[offset++];
                        if (offset + childEnumValueLen > data.Length)
                            throw new MmDecodeException("Tag child enum value data overflow");
                        tag.ChildEnum.Add(System.Text.Encoding.UTF8.GetString(data, offset, childEnumValueLen));
                        offset += childEnumValueLen;
                    }
                    break;

                default:
                    // 忽略未知的标签字段
                    break;
            }
        }

        return tag;
    }
}

File: mm-php/src/io/metamessage/mm/TagFieldParser.php
<?php

namespace io\metamessage\mm;

class TagFieldParser {

    public static function parse(array &$data, int &$offset): ?MmTag {
        $t = MmTag::empty();
        $maxOffset = count($data);
        $maxFieldCount = 255;
        $fieldCount = 0;

        while ($offset < $maxOffset) {
            if ($fieldCount++ > $maxFieldCount) {
                throw new MmDecodeException('Tag fields overflow');
            }

            $b = $data[$offset++];
            $k = $b >> 3;

            switch ($k) {
                case MmTag::K_IS_NULL:
                    $t->isNull = true;
                    break;
                case MmTag::K_EXAMPLE:
                    $t->example = true;
                    break;
                case MmTag::K_DESC:
                    $t->desc = self::readSizedString($data, $offset, $b);
                    break;
                case MmTag::K_TYPE:
                    $t->type = ValueType::fromCode($data[$offset++]);
                    break;
                case MmTag::K_RAW:
                    $t->raw = true;
                    break;
                case MmTag::K_NULLABLE:
                    $t->nullable = true;
                    break;
                case MmTag::K_ALLOW_EMPTY:
                    $t->allowEmpty = true;
                    break;
                case MmTag::K_UNIQUE:
                    $t->unique = true;
                    break;
                case MmTag::K_DEFAULT:
                    $t->defaultValue = self::readShortString($data, $offset, $b);
                    break;
                case MmTag::K_MIN:
                    $t->min = self::readShortString($data, $offset, $b);
                    break;
                case MmTag::K_MAX:
                    $t->max = self::readShortString($data, $offset, $b);
                    break;
                case MmTag::K_SIZE:
                    $t->size = self::readUint64($data, $offset, $b);
                    break;
                case MmTag::K_ENUM:
                    $t->enumValues = self::readSizedString($data, $offset, $b);
                    $t->type = ValueType::ENUM;
                    break;
                case MmTag::K_PATTERN:
                    $t->pattern = self::readShortString($data, $offset, $b);
                    break;
                case MmTag::K_LOCATION:
                    $n = $b & 7;
                    $t->locationHours = (int)self::readString($data, $offset, $n);
                    break;
                case MmTag::K_VERSION:
                    $t->version = self::readUint64($data, $offset, $b);
                    break;
                case MmTag::K_MIME:
                    $n = $b & 7;
                    if ($n < 7) {
                        $t->mime = MimeWire::toString($n);
                    } else {
                        $t->mime = MimeWire::toString($data[$offset++]);
                    }
                    break;
                case MmTag::K_CHILD_DESC:
                    $t->childDesc = self::readSizedString($data, $offset, $b);
                    break;
                case MmTag::K_CHILD_TYPE:
                    $t->childType = ValueType::fromCode($data[$offset++]);
                    break;
                case MmTag::K_CHILD_RAW:
                    $t->childRaw = true;
                    break;
                case MmTag::K_CHILD_NULLABLE:
                    $t->childNullable = true;
                    break;
                case MmTag::K_CHILD_ALLOW_EMPTY:
                    $t->childAllowEmpty = true;
                    break;
                case MmTag::K_CHILD_UNIQUE:
                    $t->childUnique = true;
                    break;
                case MmTag::K_CHILD_DEFAULT:
                    $t->childDefault = self::readShortString($data, $offset, $b);
                    break;
                case MmTag::K_CHILD_MIN:
                    $t->childMin = self::readShortString($data, $offset, $b);
                    break;
                case MmTag::K_CHILD_MAX:
                    $t->childMax = self::readShortString($data, $offset, $b);
                    break;
                case MmTag::K_CHILD_SIZE:
                    $t->childSize = self::readUint64($data, $offset, $b);
                    break;
                case MmTag::K_CHILD_ENUM:
                    $t->childEnum = self::readSizedString($data, $offset, $b);
                    $t->childType = ValueType::ENUM;
                    break;
                case MmTag::K_CHILD_PATTERN:
                    $t->childPattern = self::readShortString($data, $offset, $b);
                    break;
                case MmTag::K_CHILD_LOCATION:
                    $n = $b & 7;
                    $t->childLocationHours = (int)self::readString($data, $offset, $n);
                    break;
                case MmTag::K_CHILD_VERSION:
                    $t->childVersion = self::readUint64($data, $offset, $b);
                    break;
                case MmTag::K_CHILD_MIME:
                    $n = $b & 7;
                    if ($n < 7) {
                        $t->childMime = MimeWire::toString($n);
                    } else {
                        $t->childMime = MimeWire::toString($data[$offset++]);
                    }
                    break;
                default:
                    // 对于未知的标签字段，我们已经在循环开始时增加了 $offset
                    // 所以这里不需要再增加，直接跳过
                    break;
            }
        }

        return $t;
    }

    private static function readString(array &$data, int &$offset, int $n): string {
        $s = '';
        for ($i = 0; $i < $n; $i++) {
            $s .= chr($data[$offset++]);
        }
        return $s;
    }

    private static function readSizedString(array &$data, int &$offset, int $b): string {
        $n = $b & 7;
        if ($n <= 5) {
            return self::readString($data, $offset, $n);
        } elseif ($n == 6) {
            $l = $data[$offset++];
            return self::readString($data, $offset, $l);
        } else {
            $l = ($data[$offset++] << 8) | $data[$offset++];
            return self::readString($data, $offset, $l);
        }
    }

    private static function readShortString(array &$data, int &$offset, int $b): string {
        $n = $b & 7;
        if ($n < 7) {
            return self::readString($data, $offset, $n);
        } else {
            $l = $data[$offset++];
            return self::readString($data, $offset, $l);
        }
    }

    private static function readUint64(array &$data, int &$offset, int $b): int {
        $n = $b & 7;
        switch ($n) {
            case 0:
                return $data[$offset++];
            case 1:
                return ($data[$offset++] << 8) | $data[$offset++];
            case 2:
                return ($data[$offset++] << 16) | ($data[$offset++] << 8) | $data[$offset++];
            case 3:
                return ($data[$offset++] << 24) | ($data[$offset++] << 16) | ($data[$offset++] << 8) | $data[$offset++];
            default:
                throw new MmDecodeException('Invalid uint64 encoding');
        }
    }
}
